File: app/src/main/java/app/safeinvoice/data/excel/ExcelService.kt
package app.safeinvoice.data.excel

import android.content.Context
import android.net.Uri
import app.safeinvoice.util.Money
import org.apache.poi.hssf.usermodel.HSSFWorkbook
import org.apache.poi.ss.usermodel.CellType
import org.apache.poi.ss.usermodel.DateUtil
import org.apache.poi.ss.usermodel.Workbook
import org.apache.poi.ss.usermodel.WorkbookFactory
import org.apache.poi.xssf.usermodel.XSSFWorkbook
import java.io.BufferedReader
import java.io.File
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import java.text.SimpleDateFormat
import java.util.Locale

data class SheetPreview(
    val fileName: String,
    val sheetName: String,
    val headers: List<String>,
    val rows: List<List<String>>,
)

class ExcelService(private val context: Context) {

    fun preview(uri: Uri, displayName: String): SheetPreview {
        val name = displayName.lowercase(Locale.ROOT)
        return if (name.endsWith(".csv")) {
            val (headers, rows) = readCsv(uri)
            SheetPreview(displayName, "CSV", headers, rows.take(25))
        } else {
            context.contentResolver.openInputStream(uri)?.use { input ->
                WorkbookFactory.create(input).use { wb ->
                    val sheet = wb.getSheetAt(0)
                    val headers = rowValues(sheet.getRow(0))
                    val rows = (1..minOf(sheet.lastRowNum, 25)).map { r ->
                        rowValues(sheet.getRow(r), headers.size)
                    }
                    SheetPreview(displayName, sheet.sheetName ?: "Sheet1", headers, rows)
                }
            } ?: SheetPreview(displayName, "", emptyList(), emptyList())
        }
    }

    fun readAll(uri: Uri, displayName: String): Pair<List<String>, List<List<String>>> {
        val name = displayName.lowercase(Locale.ROOT)
        return if (name.endsWith(".csv")) {
            readCsv(uri)
        } else {
            context.contentResolver.openInputStream(uri)?.use { input ->
                WorkbookFactory.create(input).use { wb ->
                    val sheet = wb.getSheetAt(0)
                    val headers = rowValues(sheet.getRow(0))
                    val rows = (1..sheet.lastRowNum).map { r ->
                        rowValues(sheet.getRow(r), headers.size)
                    }.filter { row -> row.any { it.isNotBlank() } }
                    headers to rows
                }
            } ?: (emptyList<String>() to emptyList())
        }
    }

    fun writeWorkbook(
        dest: File,
        sheets: Map<String, Pair<List<String>, List<List<String>>>>,
        xls: Boolean = false,
    ): File {
        dest.parentFile?.mkdirs()
        val wb: Workbook = if (xls) HSSFWorkbook() else XSSFWorkbook()
        sheets.forEach { (name, content) ->
            val sheet = wb.createSheet(name.take(31))
            val headerRow = sheet.createRow(0)
            content.first.forEachIndexed { i, h -> headerRow.createCell(i).setCellValue(h) }
            content.second.forEachIndexed { r, row ->
                val excelRow = sheet.createRow(r + 1)
                row.forEachIndexed { c, v -> excelRow.createCell(c).setCellValue(v) }
            }
            content.first.indices.forEach { sheet.setColumnWidth(it, 18 * 256) }
        }
        dest.outputStream().use { wb.write(it) }
        wb.close()
        return dest
    }

    fun writeCsv(dest: File, headers: List<String>, rows: List<List<String>>): File {
        dest.parentFile?.mkdirs()
        dest.writeText(buildString {
            appendLine(headers.joinToString(",") { csvEscape(it) })
            rows.forEach { row ->
                appendLine(row.joinToString(",") { csvEscape(it) })
            }
        }, StandardCharsets.UTF_8)
        return dest
    }

    fun cell(row: List<String>, mapping: Map<String, Int>, key: String): String {
        val idx = mapping[key] ?: return ""
        if (idx < 0 || idx >= row.size) return ""
        return row[idx].trim()
    }

    fun parseAmount(raw: String): Double {
        val cleaned = raw.replace("R", "", ignoreCase = true)
            .replace("ZAR", "", ignoreCase = true)
            .replace(",", "")
            .replace("\\s".toRegex(), "")
            .trim()
        return cleaned.toDoubleOrNull()?.let { Money.round(it) } ?: 0.0
    }

    private fun readCsv(uri: Uri): Pair<List<String>, List<List<String>>> {
        val lines = context.contentResolver.openInputStream(uri)?.use { input ->
            BufferedReader(InputStreamReader(input, StandardCharsets.UTF_8)).readLines()
        } ?: return emptyList<String>() to emptyList()
        if (lines.isEmpty()) return emptyList<String>() to emptyList()
        val headers = splitCsv(lines.first())
        val rows = lines.drop(1).map { splitCsv(it, headers.size) }.filter { row -> row.any { it.isNotBlank() } }
        return headers to rows
    }

    private fun rowValues(row: org.apache.poi.ss.usermodel.Row?, minSize: Int = 0): List<String> {
        if (row == null) return List(minSize) { "" }
        val last = maxOf(row.lastCellNum.toInt(), minSize)
        return (0 until last).map { i ->
            val cell = row.getCell(i) ?: return@map ""
            when (cell.cellType) {
                CellType.STRING -> cell.stringCellValue.orEmpty()
                CellType.BOOLEAN -> cell.booleanCellValue.toString()
                CellType.NUMERIC -> {
                    if (DateUtil.isCellDateFormatted(cell)) {
                        SimpleDateFormat("yyyy-MM-dd", Locale.US).format(cell.dateCellValue)
                    } else {
                        val n = cell.numericCellValue
                        if (n % 1.0 == 0.0) n.toLong().toString() else n.toString()
                    }
                }
                CellType.FORMULA -> try {
                    cell.stringCellValue
                } catch (_: Exception) {
                    cell.numericCellValue.toString()
                }
                else -> ""
            }
        }
    }

    private fun splitCsv(line: String, minSize: Int = 0): List<String> {
        val out = mutableListOf<String>()
        val cur = StringBuilder()
        var inQuotes = false
        var i = 0
        while (i < line.length) {
            val c = line[i]
            when {
                c == '"' && inQuotes && i + 1 < line.length && line[i + 1] == '"' -> {
                    cur.append('"'); i++
                }
                c == '"' -> inQuotes = !inQuotes
                c == ',' && !inQuotes -> {
                    out += cur.toString(); cur.clear()
                }
                else -> cur.append(c)
            }
            i++
        }
        out += cur.toString()
        while (out.size < minSize) out += ""
        return out
    }

    private fun csvEscape(value: String): String {
        return if (value.contains(',') || value.contains('"') || value.contains('\n')) {
            "\"${value.replace("\"", "\"\"")}\""
        } else value
    }
}

object CustomerImportFields {
    val fields = listOf(
        "name" to "Customer name",
        "contactName" to "Contact",
        "email" to "Email",
        "phone" to "Phone",
        "addressLine1" to "Address",
        "city" to "City",
        "province" to "Province",
        "postalCode" to "Postal code",
        "vatNumber" to "VAT number",
        "notes" to "Notes",
    )
}

object InvoiceImportFields {
    val fields = listOf(
        "customerName" to "Customer name",
        "number" to "Invoice number",
        "status" to "Status",
        "issueDate" to "Issue date",
        "dueDate" to "Due date",
        "description" to "Line description",
        "quantity" to "Quantity",
        "unitPrice" to "Unit price",
        "total" to "Total",
        "notes" to "Notes",
    )
}

fun guessMapping(headers: List<String>, fields: List<Pair<String, String>>): Map<String, Int> {
    val normalized = headers.map { it.lowercase(Locale.ROOT).replace(Regex("[^a-z0-9]"), "") }
    return fields.associate { (key, label) ->
        val aliases = listOf(key, label).map { it.lowercase(Locale.ROOT).replace(Regex("[^a-z0-9]"), "") }
        val idx = normalized.indexOfFirst { h -> aliases.any { a -> h.contains(a) || a.contains(h) } }
        key to idx
    }
}

File: lib/data/excel_service.dart
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:excel/excel.dart';
import 'package:spreadsheet_decoder/spreadsheet_decoder.dart';

import '../domain/money.dart';

class SheetPreview {
  const SheetPreview({
    required this.fileName,
    required this.sheetName,
    required this.headers,
    required this.rows,
  });

  final String fileName;
  final String sheetName;
  final List<String> headers;
  final List<List<String>> rows;
}

class ExcelService {
  SheetPreview preview(Uint8List bytes, String displayName, {int maxRows = 25}) {
    final all = readAll(bytes, displayName);
    return SheetPreview(
      fileName: displayName,
      sheetName: all.$1,
      headers: all.$2,
      rows: all.$3.take(maxRows).toList(),
    );
  }

  (String sheetName, List<String> headers, List<List<String>> rows) readAll(
    Uint8List bytes,
    String displayName,
  ) {
    final name = displayName.toLowerCase();
    if (name.endsWith('.csv') || _looksLikeCsv(bytes)) {
      final parsed = _readCsv(utf8.decode(bytes, allowMalformed: true));
      return ('CSV', parsed.$1, parsed.$2);
    }
    try {
      final decoder = SpreadsheetDecoder.decodeBytes(bytes, update: false);
      final table = decoder.tables.values.first;
      final raw = table.rows
          .map((r) => r.map((c) => _cellString(c)).toList())
          .where((r) => r.any((c) => c.trim().isNotEmpty))
          .toList();
      if (raw.isEmpty) return (table.name, <String>[], <List<String>>[]);
      final headers = raw.first;
      final width = headers.length;
      final rows = raw.skip(1).map((r) {
        final copy = List<String>.from(r);
        while (copy.length < width) {
          copy.add('');
        }
        return copy.take(width).toList();
      }).toList();
      return (table.name, headers, rows);
    } catch (_) {
      final parsed = _readCsv(utf8.decode(bytes, allowMalformed: true));
      return ('CSV', parsed.$1, parsed.$2);
    }
  }

  File writeWorkbook(
    File dest,
    Map<String, (List<String>, List<List<String>>)> sheets,
  ) {
    dest.parent.createSync(recursive: true);
    final excel = Excel.createExcel();
    final defaultSheet = excel.getDefaultSheet();
    var first = true;
    sheets.forEach((name, content) {
      final sheetName = name.length > 31 ? name.substring(0, 31) : name;
      final sheet = excel[sheetName];
      for (var c = 0; c < content.$1.length; c++) {
        sheet.cell(CellIndex.indexByColumnRow(columnIndex: c, rowIndex: 0)).value =
            TextCellValue(content.$1[c]);
      }
      for (var r = 0; r < content.$2.length; r++) {
        for (var c = 0; c < content.$2[r].length; c++) {
          sheet
              .cell(CellIndex.indexByColumnRow(columnIndex: c, rowIndex: r + 1))
              .value = TextCellValue(content.$2[r][c]);
        }
      }
      if (first && defaultSheet != null && defaultSheet != name) {
        excel.delete(defaultSheet);
        first = false;
      } else {
        first = false;
      }
    });
    dest.writeAsBytesSync(excel.encode() ?? <int>[]);
    return dest;
  }

  File writeCsv(File dest, List<String> headers, List<List<String>> rows) {
    dest.parent.createSync(recursive: true);
    final buf = StringBuffer();
    buf.writeln(headers.map(_csvEscape).join(','));
    for (final row in rows) {
      buf.writeln(row.map(_csvEscape).join(','));
    }
    dest.writeAsStringSync(buf.toString());
    return dest;
  }

  String cell(List<String> row, Map<String, int> mapping, String key) {
    final idx = mapping[key];
    if (idx == null || idx < 0 || idx >= row.length) return '';
    return row[idx].trim();
  }

  double parseAmount(String raw) {
    final cleaned = raw
        .replaceAll(RegExp(r'R', caseSensitive: false), '')
        .replaceAll(RegExp(r'ZAR', caseSensitive: false), '')
        .replaceAll(',', '')
        .replaceAll(RegExp(r'\s'), '')
        .trim();
    return Money.round(double.tryParse(cleaned) ?? 0);
  }

  bool _looksLikeCsv(Uint8List bytes) {
    final probe = utf8.decode(bytes.take(200).toList(), allowMalformed: true);
    return probe.contains(',') && !probe.contains('\u0000');
  }

  (List<String>, List<List<String>>) _readCsv(String text) {
    final lines = const LineSplitter().convert(text).where((l) => l.trim().isNotEmpty).toList();
    if (lines.isEmpty) return (<String>[], <List<String>>[]);
    final headers = _splitCsv(lines.first);
    final rows = lines.skip(1).map((l) => _splitCsv(l, headers.length)).toList();
    return (headers, rows);
  }

  List<String> _splitCsv(String line, [int minSize = 0]) {
    final out = <String>[];
    final cur = StringBuffer();
    var inQuotes = false;
    for (var i = 0; i < line.length; i++) {
      final c = line[i];
      if (c == '"' && inQuotes && i + 1 < line.length && line[i + 1] == '"') {
        cur.write('"');
        i++;
      } else if (c == '"') {
        inQuotes = !inQuotes;
      } else if (c == ',' && !inQuotes) {
        out.add(cur.toString());
        cur.clear();
      } else {
        cur.write(c);
      }
    }
    out.add(cur.toString());
    while (out.length < minSize) {
      out.add('');
    }
    return out;
  }

  String _csvEscape(String value) {
    if (value.contains(',') || value.contains('"') || value.contains('\n')) {
      return '"${value.replaceAll('"', '""')}"';
    }
    return value;
  }

  String _cellString(Object? cell) {
    if (cell == null) return '';
    if (cell is DateTime) {
      return '${cell.year.toString().padLeft(4, '0')}-'
          '${cell.month.toString().padLeft(2, '0')}-'
          '${cell.day.toString().padLeft(2, '0')}';
    }
    if (cell is num) {
      if (cell % 1 == 0) return cell.toInt().toString();
      return cell.toString();
    }
    return cell.toString();
  }
}

Map<String, int> guessMapping(List<String> headers, List<(String, String)> fields) {
  final normalized = headers
      .map((h) => h.toLowerCase().replaceAll(RegExp(r'[^a-z0-9]'), ''))
      .toList();
  const extraAliases = {
    'quantity': ['qty', 'qnty', 'amountqty'],
    'unitPrice': ['price', 'rate', 'unitcost'],
    'customerName': ['client', 'clientname', 'customer'],
    'vatNumber': ['vat', 'taxnumber', 'vatno'],
    'postalCode': ['postcode', 'zip', 'zipcode'],
    'contactName': ['contact', 'attn', 'attention'],
    'addressLine1': ['address', 'street'],
    'name': ['customername', 'client', 'company'],
  };
  return {
    for (final field in fields)
      field.$1: () {
        final aliases = [field.$1, field.$2, ...?extraAliases[field.$1]]
            .map((a) => a.toLowerCase().replaceAll(RegExp(r'[^a-z0-9]'), ''))
            .toList();
        return normalized.indexWhere(
          (h) =>
              h.isNotEmpty &&
              aliases.any((a) => a.isNotEmpty && (h == a || h.contains(a) || a.contains(h))),
        );
      }(),
  };
}
